src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs

```diff
@@ -689,6 +689,114 @@ protected override void ProcessRecord()
         }
 
         #endregion Overrides
+
+        /// <summary>
+        /// Adds UserName, Description, BinaryPathName, DelayedAutoStart and StartupType to a ServiceController object.
+        /// </summary>
+        /// <param name="service"></param>
+        /// <returns>ServiceController as PSObject with UserName, Description and StartupType added</returns>
+        private PSObject AddProperties(ServiceController service) {
+            NakedWin32Handle hScManager = IntPtr.Zero;
+            NakedWin32Handle hService = IntPtr.Zero;
+            int lastError = 0;
+            PSObject serviceAsPSObj = PSObject.AsPSObject(service);
+            IntPtr descriptionStructPtr = IntPtr.Zero;
+            IntPtr delayedAutoStartStuctPtr = IntPtr.Zero;
+            IntPtr serviceConfigStructPtr = IntPtr.Zero;
+            try
+            {
+                hScManager = NativeMethods.OpenSCManagerW(
+                    lpMachineName: service.MachineName,
+                    lpDatabaseName: null,
+                    dwDesiredAccess: NativeMethods.SC_MANAGER_CONNECT
+                );
+                if (IntPtr.Zero == hScManager) {
+                    lastError = Marshal.GetLastWin32Error();
+                    Win32Exception exception = new Win32Exception(lastError);
+                    WriteNonTerminatingError(
+                        service,
+                        service.MachineName,
+                        exception,
+                        "ComputerAccessDenied",
+                        ServiceResources.ComputerAccessDenied,
+                        ErrorCategory.PermissionDenied);
+                }
+                hService = NativeMethods.OpenServiceW(
+                    hScManager,
+                    service.ServiceName,
+                    NativeMethods.SERVICE_QUERY_CONFIG
+                );
+                if (IntPtr.Zero == hService) {
+                    lastError = Marshal.GetLastWin32Error();
+                    Win32Exception exception = new Win32Exception(lastError);
+                    WriteNonTerminatingError(
+                        service,
+                        exception,
+                        "CouldNotGetServiceInfo",
+                        ServiceResources.CouldNotGetServiceInfo,
+                        ErrorCategory.PermissionDenied);
+                }
+
+                bool querySuccessful = NativeMethods.QueryServiceConfig2(hService, NativeMethods.SERVICE_CONFIG_DESCRIPTION, out descriptionStructPtr);
+                NativeMethods.SERVICE_DESCRIPTIONW description = (NativeMethods.SERVICE_DESCRIPTIONW)Marshal.PtrToStructure(descriptionStructPtr, typeof(NativeMethods.SERVICE_DESCRIPTIONW));
```

Comment:
@joandrsn Could you please address @mirichmo previous request:

I'd like to see these method calls organized differently to protect against potential memory leaks. The pattern is repeated in NativeMethods.QueryServiceConfig2 and I'd like all instances of it fixed. Since NativeMethods.QueryServiceConfig() is a helper function, its output parameter, instead of being an IntPtr structurePointer, should be the managed object NativeMethods.SERVICE_DESCRIPTIONW. That way, the unmanaged memory that is allocated within that function is also cleaned up within that function. The managed object carries the values out of the function and eventually gets garbage collected. To fix this instance, line 741 should be moved within the helper function. You can probably make Nativemethods.QueryServiceConfig2() a generic method as well to avoid having to use a switch for infoLevel.